add the proposal draft as a change (#1)
dietmarkuehl/P3552-task: lazy.md

`void` for coroutines not producing any value. For example:

    int main() {
        ex::sync_wait([]->ex::lazy<>{
            int result = co_await []->ex::lazy<int> { co_return 42; }();
            assert(result == 42);
        }());
    }

The inner coroutines completes with `set_value_t(int)` which gets
translated to the value returned from `co_await` (see [`co_await`
result type below](#result-type-for-co_await) for more details).
The outer coroutine completes with `set_value_t()`.

Beyond the result type there are a number of features for a coroutine
task which benefit from customisation or for which it may be desirable
to disable them because they introduce a cost. As many template
parameters become unwieldy, it makes sense to combine these into a
[defaulted] context parameter. The aspects which benefit from
customisation are at least:

- [Customising the environment](#environment-support) for child
    operations. The context itself can actually become part of
    the environment.
- Disable [scheduler affinity](#scheduler-affinity) and/or configure
    the strategy for obtaining the coroutine's scheduler.
- Configure [allocator awareness](#allocator-support).

> - Indicate that the coroutine should be [noexcept](#error-reporting).

Reviewer: I assume the ship has sailed on making `noexcept` do something useful on coroutine function declarations?

Author: That is unrelated to any given coroutine class! I think, `noexcept` on the coroutine [factory] function would indicate that the coroutine doesn't throw until the first real suspension point is reached. If there would be desire to do something better, it would be a Core paper, I think.

Reviewer: Right! My point is that with `std::lazy` adding `noexcept` to the function signature is tatamount to saying "terminates if allocating the coro frame fails" or something very similar. It's not an assertion about any of the code in the coroutine function itself. So, this current proposal shows a bit of a wart in the language spec.

I actually suspect the Right UX is to ban `noexcept` on coro functions, and explicitly require either `noexcept-construct` or `noexcept-await`. And yes, I agree this is a separate proposal, it merely related to the issues discussed here.

[P1056r0](https://wg21.link/P1056r0) the name was `task`; the primary
change for [P1056r1](https://wg21.link/P1056r1) is changing the
name to `lazy`). The fundamental idea is to have a coroutine type
which can be `co_await`ed: the interface of `lazy` consists of move
constructor, deliberately no move assignment, a destructor, and
`operator co_await()`. The proposals doesn't go into much detail
on how to eventually use a coroutine but mentions that there could
be functions like `sync_await(task<To>)` to await completion of a
task (similar to
[`execution::sync_wait(sender)`](https://eel.is/c++draft/exec.sync.wait))
or a few variations of that.

A fair part of the paper argues why `future.then()` is _not_ a good
approach to model coroutines and their results. Using `future`
requires allocation, synchronisation, reference counting, and
scheduling which can all be avoided when using coroutines in a
structured way.

The paper also mentions support for [symmetric
transfer](https://wg21.link/P0913) and allocator support. Both of these
are details on how the coroutine is implemented.

[Discussion for P1056r0 in SG1](https://wiki.edg.com/bin/view/Wg21rapperswil2018/P1056R0)

> - The task doesn't really have to anything with concurrency.

Reviewer 2: ```suggestion
- The task doesn't really have anything to do with concurrency.
```

there isn't a need to support a separate `shared_task<T>` in a
sender/receiver world. Likewise, throwing of results can be avoid
by suitably rewriting the result of the `set_error` channel avoiding
the need for an operation akin to `when_ready()`.

## [libunifex](https://github.com/facebookexperimental/libunifex)

`libunifex` is an earlier implementation of the sender/receiver
ideas. Compared to `std::execution` it is lacking some of the
flexibilities. For example, it doesn't have a concept of environments
or domains. However, the fundamental idea of three completion
channels for success, failure, and cancellation and the general
shape of how these are used is present (even using the same names
for `set_value` and `set_error`; the equivalent of `set_stopped`
is called `set_done`). `libunifex` is in production use in multiple
places. The implementation includes a
[`unifex::task<T>`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/task.hpp).

As `libunifex` is sender/receiver-based, its `unifex::task<T>` is
implemented such that `co_await` can deal with senders in addition
to awaitables or awaiters. Also, `unifex::task<T>` is _scheduler

> affine_: the coroutine code resume on the same scheduler even if a

Reviewer 2: ```suggestion
affine_: the coroutine code resumes on the same scheduler even if a
```

sender completed on a different scheduler. The corresponding scheduler
is taken from the receiver.  The exception for this rule is when
explicitly awaiting the result of `scheduler(scheduler)`.  The
relevant treatment is in the promise type's `await_transform()`:

- If a sender `s` which is result of `schedule(scheduler)` is
  `co_await`ed, the corresponding `scheduler` is installed as the
  task's scheduler and the task resume on the result of completing

>   `s`.

Reviewer 2: For what it's worth, I regret this choice. I think it's important for a task to be able to deliberately switch its current scheduler, but that ability should be made available through a more explicit interface rather than making `co_await schedule(s);` magical.

Author: I'm certainly not found of making `co_await schedule(s)` magical because I don't see how it would be possible to determine for sender if it came from a `schedule(s)` operation. It is also a bit subtle. That said, I'm also a bit lost at what would be possible alternatives. I can imagine a different spelling than `scheduler(s)`, e.g., `transfer_task_to(s)` which could be something different than a sender as it is distinctly intended to interact with the `promise_type`'s `await_transform`. I do think the transfer is asynchronous and will need to be awaited, though.

additional allocation for each `co_await`.

# Objectives

Also see [sender/receiver issue 241](https://github.com/cplusplus/sender-receiver/issues/241).

Based on the prior work and discussions around corresponding coroutine
support there a number of required or desired features (listed in
no particular order):

1. A coroutine task needs to be awaiter/awaitable friendly. While
    that seems obvious, it is, e.g., possible to create an
    `await_transform` which is deleted for awaiters.
2. For sender algorithms it is common to adapt the results using
    suitable algorithms. When awaiting sender in a coroutine it may
    be annoying having to transform the result into a shape which
    is friendly to a coroutine use. Thus, it may be reasonable to
    support rewriting certain shapes of completion signatures into
    something different to make the use of senders easier in a
    coroutine task.
3. A coroutine task needs to be sender friendly: it is expected that
    asynchronous code is often written using coroutines awaiting
    senders. However, it isn't as obvious as not all senders will
    necessarily be awaitable. For example neither unifex nor stdexec

>     support senders with more than `set_value` completion.

Reviewer 3: ```suggestion
    support senders with more than one `set_value` completion.
```

4. It is possibly confusing and problematic if coroutines resume on
    a different execution context than the one they were suspended
    on.  Senders could, however, complete on an entirely different
    scheduler than where they started. As a result, a coroutine
    task should be scheduler affine by default, i.e., it should
    always resume on the same scheduler. There should probably be
    a way to opt out of scheduler affinity when the implications are
    well understood.
5. When using coroutines there will probably be an allocation at
    least for the coroutine frame. To support the use in environments
    where memory allocations using `new`/`delete` aren't supported
    the coroutine task should support allocations using allocators.
6. Receivers have associated environments with attributes which are

>     generally forwarded. It should be possible to provide provide a

Reviewer 3: ```suggestion
    generally forwarded. It should be possible to provide a
```

>     user-customisable environment from the receiver used by `co_await`

Reviewer 3: I know what you want to say: we should be able to specify the amount of queries that are forwarded to awaitables within our coroutine. But I didn't understand it from the text. Maybe I'm wrong?

Author: OK. I'll try to clarify.

Reviewer 3: Better

places. The implementation includes a
[`unifex::task<T>`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/task.hpp).

As `libunifex` is sender/receiver-based, its `unifex::task<T>` is
implemented such that `co_await` can deal with senders in addition
to awaitables or awaiters. Also, `unifex::task<T>` is _scheduler
affine_: the coroutine code resume on the same scheduler even if a
sender completed on a different scheduler. The corresponding scheduler
is taken from the receiver.  The exception for this rule is when
explicitly awaiting the result of `scheduler(scheduler)`.  The
relevant treatment is in the promise type's `await_transform()`:

- If a sender `s` which is result of `schedule(scheduler)` is
  `co_await`ed, the corresponding `scheduler` is installed as the
  task's scheduler and the task resume on the result of completing
  `s`.
- For both senders and awaiters being awaited, the coroutine
  will be resumed on the task's current scheduler when the task is
  scheduler affine unless the sender algorithm is tagged to inhibit
  transfer to the current scheduler (using a `static constexpr`
  member named `blocking` which is initialised to

>   `blocking_kind::always_inline`).

Reviewer 2: This bullet is ambiguous to me. On first read, I think you're saying that, with the appropriate incantation, it's possible to cause a `unifex::task<>` to not be scheduler affine, but that's not true. On second read, I wonder if you're trying to say that, when awaiting a sender from within a task, if that sender is marked as `always_inline` then rescheduling isn't needed upon coroutine resumption because we're already on the right execution context; _that's_ true, but it's not super clear in the current wording.

Author: I meant that there is no scheduling operation when senders are marked as `always_inline`. I will clarify.

- If a sender `s` is `co_await`ed it gets `connect`ed to a receiver
  provided by the task to form an awaiter holding the operation
  state which gets `start`ed by the awaiter's `await_suspend`. The
  receiver arranges for the `set_value` completion to become a
  result return from `await_resume`, a `set_error` completon to
  become an exception, and a `set_done` completion to not resume

>   the task.

Reviewer 2: ```suggestion
  become an exception, and a `set_done` completion to resume an
  a special "on done" coroutine handle rather than resuming the task
  itself.
```

When `co_await`ing a sender `s` there can be at most one `set_value`
completion: if there is more than one `set_value` completion the
`await_transform` will do nothing and the result sending cannot be
`co_await`ed (unless it is also given an awaitable interface). The
result type of `co_await s` depends on the number of arguments to
`set_value`:

> - If there is no for `set_value`  then the type of `co_await s`

Reviewer 2: ```suggestion
- If there are no arguments for `set_value`  then the type of `co_await s`
```

`unifex::async_scope s` it isn't possible to directly call `s.spawn(t)`
with a `unifex::task<T> t` as the `async_scope` doesn't provide a
scheduler.

`libunifex` provides some sender algorithms to transform the sender
result into something which may be more suitable to be `co_await`ed.
For example, `unifex::done_as_optional(sender)` turns a successful
completion for a type `T` into an `std::optional<T>` and the
cancellation completion `set_done` into a `set_value` completion
with a disengaged `std::optional<T>`.

The `unifex::task<T>` is itself a sender and can be used correspondingly.
To deal with scheduler affinity a type erase scheduler
`unifex::any_scheduler` is used.

The `unifex::task<T>` doesn't have allocator support. When creating
a task multiple objects are allocated on the heap: it seems there
is a total of 6 allocations for each `unifex::task<T>` being created.
After that, it seem the different `co_await`s don't use a separate
allocation.

The `unifex::task<T>` doesn't guard against stack overflow directly
guard against stack overflow. Due to rescheduling continuations on
a scheduler when the completion isn't always inlne, the issue only

> arises when `co_await`ing many senders with `blocking_kind::always_inline`.

Reviewer 2: ```suggestion
The `unifex::task<T>` doesn't directly guard against stack overflow.
Due to rescheduling continuations on a scheduler when the completion
isn't always inline, the issue only arises when `co_await`ing many
senders with `blocking_kind::always_inline`.
```

    completes inline.
- Unlike unifex, it is OK if the receiver's environment doesn't provide
    a scheduler. In that case an inline scheduler is used. If an inline
    scheduler is used there is the possibility of stack overflow.
- It is possible to `co_await just_error(e)` and `co_await just_stopped()`,
    i.e., the sender isn't required to have a `set_value_t` completion.

The `exec::task<T, C>` also provides a _context_ `C`. An object of
this type becomes the environment for receivers `connect()`ed to
`co_await`ed senders. The default context provides access to the
task's scheduler. In addition an `in_place_stop_token` is provides
which forwards the stop requests from the environment of the receiver
which is connected to the task.

Like the unifex task `exec::task<T, C>` doesn't provide any allocator
support. When creating a task there are two allocations and an
additional allocation for each `co_await`.

# Objectives

Also see [sender/receiver issue 241](https://github.com/cplusplus/sender-receiver/issues/241).

Based on the prior work and discussions around corresponding coroutine
support there a number of required or desired features (listed in
no particular order):

> 1. A coroutine task needs to be awaiter/awaitable friendly. While

Reviewer 2: I'm not 100% certain I know what "awaiter/awaitable friendly" means; are you saying that it must be possible to await a "normal" awaitable that isn't a sender?

Author: In theory a `promise_type` can have an awaiter/awaitable hostile `await_resume`, making it `= delete`. I agree that doing so would be an odd choice but technically it is possible. So, yes, I'm saying that it should be possible to await a normal awaitable. Note that `execution::sender<T>` does consider awaitables senders and awaitables are treated accordingly, e.g., by `execution::connect`.

    that seems obvious, it is, e.g., possible to create an
    `await_transform` which is deleted for awaiters.
2. For sender algorithms it is common to adapt the results using

>     suitable algorithms. When awaiting sender in a coroutine it may

Reviewer 2: This first sentence is unclear to me; are you talking specifically about awaiting senders from within a coroutine, or more generally about the fact that it's common to compose senders together?

Author: I'll write that to hopefully make it clear that senders may be used differently when composing them without coroutines and when using them in coroutines. Effectively, this point is intended to address the concern brought up during the discussion on concurrent queue in Wroclaw that using something akin to `co_await into_optional(q.async_pop())` is unreasonable (I don't think that is necessarily a majority position, though).

    be annoying having to transform the result into a shape which
    is friendly to a coroutine use. Thus, it may be reasonable to
    support rewriting certain shapes of completion signatures into
    something different to make the use of senders easier in a
    coroutine task.
3. A coroutine task needs to be sender friendly: it is expected that
    asynchronous code is often written using coroutines awaiting
    senders. However, it isn't as obvious as not all senders will

>     necessarily be awaitable. For example neither unifex nor stdexec

Reviewer 2: "…it isn't as obvious…" what is not as obvious as what?

Author: It isn't obvious, that all senders can necessarily be `co_await`ed. While that was in my head, the text doesn't really express that. I'll try to clarify. The intent here is to lead up to an extension beyond what unifex and stdexec can `co_await` although the result may be somewhat unwieldy (i.e., a `variant` of `tuple`s).

    support senders with more than one `set_value` completion.
4. It is possibly confusing and problematic if coroutines resume on
    a different execution context than the one they were suspended

>     on.  Senders could, however, complete on an entirely different

Reviewer 2: I agree with this, but I think it's not obvious why I expect senders to be non-affine and coroutines to be affine. I think the reason comes down to the textual similarity between coroutine bodies and regular functions, and it'd be pretty weird if a function body started on one execution context and finished elsewhere without some big flashing lights explaining how and why that happened. In contrast, senders are a whole new thing, and there are some senders (like, say, `continue_on`) that exist simply to change execution context mid-pipeline.

Given the non-obviousness of this asymmetry between coroutines and senders, perhaps it's worth elaborating on the _why_ somewhere; if you think the why doesn't fit right here, a forward reference to where it will be explored later in the doc would be useful.

Author: I added a bit more explanation to the text.

    scheduler than where they started. As a result, a coroutine
    task should be scheduler affine by default, i.e., it should
    always resume on the same scheduler. There should probably be
    a way to opt out of scheduler affinity when the implications are
    well understood.
5. When using coroutines there will probably be an allocation at
    least for the coroutine frame. To support the use in environments
    where memory allocations using `new`/`delete` aren't supported
    the coroutine task should support allocations using allocators.

> 6. Receivers have associated environments which can support an open

Reviewer 2: I think we need a way to specify which queries a task requires of its receiver's environment and the signatures of those queries.

For example, adapting Unifex's scheduler affinity approach to `std::task<>`, the default `task` template would require that the receiver's environment have `auto get_scheduler() -> any_scheduler`, however, it might be the case that you've written a coroutine that can only run on an Nvidia GPU, in which case, you might want to override the default and say that the environment must have `auto get_scheduler() -> nvidia_scheduler`.

Furthermore, there's an existing pattern in some concurrent code of having a "request context" in TLS that contains a blob of state that represents "the current request that this thread is building the response to". [Folly has one here](https://github.com/facebook/folly/blob/main/folly/io/async/Request.h). In a sender/receiver world, custom environment queries look like they'll replace this request context business, and async operations will have direct, non-TLS access to per-operation state. In this hypothetical future, I expect coroutine authors to want to be able to specify a domain-specific environment query for, say, the user ID of the user that started the current operation. To do this, we'll need a way to express "a standard task that requires its receiver's environment to support queries for `my_company::user_id`s".

Author: Yes. That is what I'm trying to express on that bullet. I have expanded the text a bit to cover a bit more of your examples. And I totally agree that something like that is needed. Of course, it will also be seen as a form of global variables (as is TLS which should really be TLGS == thread local, global storage). I guess, at some point I should also clarify that "scheduler affinity" doesn't mean that you always carry on using the same thread: you'd get back to the same context but that may be a thread pool.

    set of queries. Normally, queries on an environment can be
    forwarded to the environment of a `connect()`ed receiver.

>     Since the coroutine types are determined before the coroutines

Reviewer 2: ```suggestion
    Since the coroutine types are determined before the coroutine's
```

    receiver is known and the queries themselves don't specify a
    result type that isn't possible when a coroutine provides a
    receiver to a sender in a `co_await` expression. It should still
    be possible to provide a user-customisable environment from the
    receiver used by `co_await` expressions. One aspect of this
    environment is to forward stop requests to `co_await`ed child
    operations.
7. The coroutine should be able to indicate that it was cancelled,
    i.e., to get `set_stopped()` called on the task's receiver.
    `std::execution::with_awaitable_senders` already provided this
    ability senders being `co_await`ed but that doesn't necessarily
    extend to the coroutine implementation.
8. Similar to indicating that a task got cancelled it would be good
    if a task could indicate that an error occurred without throwing

>     an exception which escapes from the coroutine.

Reviewer 2: I think I probably agree with this, but I don't think it's obvious. The natural way for a C++ function to report errors is by throwing something and the natural mapping from a throwing coroutine when treating that coroutine as a sender is to propagate the thrown exception through `set_error(std::exception_ptr)`.

Meta has found that throwing exception through a chain of coroutines is very expensive (it's worse than throwing through a similar chain of functions because the exception has to be caught, stored, and rethrown at each "link" in the chain) so `folly::coro::Task<>` has special handling for `co_await co_error(SomeException());` instead of `throw SomeException();`, plus various other tricks for "catching" exceptions through the value channel instead of the exception channel, but it all feels like a giant kludge. The interface on `folly::coro::Task<>` has grown quite wide to accommodate all its different use cases and I think it's _probably_ a mistake to mirror that into `std::task<>`; I suspect it would be better to provide a configurable `basic_task` template (analogous to `basic_string`) that lets the user pick and choose the features they want, plus a small collection of algorithms (such as `stopped_as_optional`, or maybe something like `yield_stopped()` or `yield_error`) for controlling what happens at various significant junctures (e.g. awaited sender cancels, awaited sender has multiple set_value completions, task wants to report an error without throwing, task wants to cancel, etc.).

Author: I'm not disagreeing that exceptions are a natural way to indicate errors. However, it is quite clear that many people disagree with that stance and rather use `std::expected<T, E>`. Also, exceptions aren't always an option: they are explicitly prohibited in some places or are unavailable, e.g., on some free-standing environments. Thus, I think providing a way which allows reporting errors without exception seems reasonable. At least, I want to offer some ways to do it. If the committee decides that these aren't needed that's fine.

Some form of configuration, probably via a [defaulted] "context" argument to `lazy<T, C>`, is probably needed. For example, the task would need to know about potential error types it needs to support. Beyond that, for me the main driver is whether a feature follows the zero cost principle and if it doesn't it should probably be configurable (scheduler affinity and allocator support both have a cost, for example). I think once the user has explicitly opted into specifying the error types to be reported, directly completing within an error without exceptions comes for free, i.e., neither does it require additional memory nor any run-time (I'll have to verify that in an implementation which isn't just a prototype for the error reporting, though).

Reviewer: @dietmarkuehl, I agree with the intent behind this bullet. I would suggest an opinionated design, though of course that would require a separate proposal. Roughly:
 - `std::result<T>` wraps `std::expected<T, std::exception_ptr>`
 - In a `std::lazy`, writing `co_await std::co_await_result(someCoro())` never throws, but returns a `std::result<RetOfCoro>`.
 
There are various niceties one gets from a `std::result` type:
 - Unlike `std::expected` it would be `[[nodiscard]]` to ensure error handling
 - Unlike `std::expected` it can be implicitly constructible from `T`, which much improves usability
 - It can be a coroutine itself, for explicit short-circuiting error handling via `co_await`, with the caveat that awaiting it from `std::lazy would require `co_await std::co_unwrap_result(someResultFn())`.
 - The `std::result` coroutine would be an automatic exception boundary.
 - Mostly non-throwing API: If `std::result` is a coroutine, normal access is either via `co_await` (in result coros) or via `co_unwrap_result` (in lazy coros). This means that instead of exposing throwing `operator*` and `value()` like `std::expected`, the only throwing accessor can be explicit: `value_or_throw()`.
 - As a counterpart to `std::unexpected`, one could expose a richer `std::error_result` that wraps `exception_ptr` with some utility methods.

I'm writing this specifically because have an implementation I could potentially move into `folly/` in the coming months. LMK if that's useful.

This direction would benefit from fixing some of the performance warts around `std::exception_ptr` as-standardized, I can easily make a laundry list based on fixes in `folly`.

Author: I _think_ I did write part of what you describe in my section on error handling: using a suitable marker type (I'm calling it `with_error` in my current implementation) and using that to indicate that the coroutine should be abandoned and `set_error(...)` should be called on the coroutine's receiver.

The section on error handling effectively recommends actually _not_ doing that (although only quite mildly) and instead using `co_yield` to create another exit path!

The description or your `error_result` is reasonable in a world where asynchronous work is always `co_await`ed. While my expectation is that using coroutines will be the primary way to interface the asynchronous interfaces from the standard library (or otherwise), they are conceptually (and in the implementation) actually using senders. For senders the error value doesn't become a return value but rather a function parameters. I don't think `[[nodiscard]]` has an effect on function parameters. The effect could be brought out when `co_await` actually result in producing such a value.

In any case, I _think_ the error handling support using exception is having something (and exceptions can be avoid by adapting the child sender) and anything beyond that is kind of new grounds.  Have a look at my section on error handling: it assumes sender/receiver, of course, but it can be entirely non-throwing, falling back to throwing if the user doesn't actually use the additional bits.

9. In general a task has to assume that an exception escapes a the
    coroutine implementation. As a result, the task's completion
    signatures need to include `set_error_t(std::exception_ptr)`.

>     If it can be indicated to the tas that no exception will escape

Reviewer 2: ```suggestion
    If it can be indicated to the task that no exception will escape
```

>     the coroutine, this completion signature can be avoided.

Reviewer 2: `unifex::task<T>` is actually an alias for `unifex::_task::_task<T, false>` where the `false` means "not `noexcept`"; there's also `unifex::nothrow_task<T>`, which is an alias for `unifex::_task::_task<T, true>`. `nothrow_task` turns exceptions thrown inside the coroutine body into calls to `std::terminate()` by analogy to throwing an exception inside a `noexcept` function. Due to a limitation in Unifex (no way to know if `connect` or `set_value` might throw), basically all senders report that they may invoke `set_error(std::exception_ptr)` but `std::execution` could avoid this.

Author: `std::execution` doesn't always include `set_error_t(exception_ptr)`. However, a subtle error quickly gets you that completion signature, e.g., if you use `then([]{})` instead of `then([]()noexcept{})`.

`set_value`, `set_error`, and `set_stopped` are required to be `noexcept` for `std::execution`. I don't know how `connect` is relevant in general: it is used prior to `start()` and if that throws it isn't something which gets reported using `set_error`. Instead, that error needs to be reported by whatever is using `connect` (of course, that in turn _may_ be a as part of executing a sender's operation state, e.g., `let_value`).

It would be nice if `lazy<T, C>` could have completion signatures which do not include `set_error_t(excpetion_ptr)` and I think we are in agreement that this would be an option and likely not the default. I imagine that it would be indicated via the context argument `C`.

10. When many `co_await`ed operations complete synchronously, there
    is a chance for stack overflow. It may be reasonable to have
    the implementation prevent that by using a suitable scheduler
    sometimes.
11. In some situations it can be useful to somehow schedule an
    asychronous clean-up operation which is triggered upon

>     coroutine exit.

Reviewer 2: This is true, but the approach taken by Folly ([`folly::coro::co_scope_exit`](https://github.com/facebook/folly/blob/main/folly/coro/ScopeExit.h#L357)) and Unifex ([`unifex::at_coroutine_exit`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/at_coroutine_exit.hpp#L386)) probably shouldn't be emulated. The lifetimes are weird and confusing, and the rules around which references can and can't be captured are hard to learn. @snarkmaster has been working on an "async closure" in Folly that will replace `co_scope_exit` and I think it's a much better approach.

Author: I carried that over from one of the issues for stdexec which listed it without any hint on how it could be addressed. I don't have a design and the proposal doesn't list anything to address that, yet. I wanted to list it for awareness in case not having a solution is show stopper for enough people.

Reviewer: The distinction between the two designs is like so:
 - `co_scope_exit` / `at_coroutine_exit`: 
   * Inside-out: guaranteed lifetime inversion since the "cleanup coro frame" outlives the locals of the coro it might reference.
   * Unfortunate syntax with lambdas returning a tuple of references into the "cleanup coro frame"
   * Requires explicit hooks in the coroutine implementation to support 
 - `async_closure`:
   * Simple wrapper, with the cleanup frame attached from outside the coro -- no lifetime issues
   * Does not require support from the coroutine type -- although it is very valuable if the coroutine type supports being wrapped, compositionally
   * Cleanup arguments look like regular coro arguments that are smart pointers 

I agree with "async RAII" or "async cleanup" being listed as part of the objectives. If this comes up in a WG discussion, I would personally say that the "best UX" is for it to be a language feature -- hence a separate proposal (high-complexity), and the "next best UX" is an `async_closure` style wrapper. IMO the inside-out cleanup scheduling of `co_scope_exit` is a last-resort feature that can paper over bad ownership designs, but I see no reason to put it into a new standard.

Author: I'll add a pointer to the discussion in the design below (Asynchronous Clean-up) where I'm already saying that the recommendation is to not do it as part of the task and rather use a different approach (currently I'm only pointing to Kirk's asynchronous object; if there is another pointer I should add to an alternative I'm happy to do so).

The algorithm `std::execution::as_awaitable` does turn a sender
into an awaitable and is expected to be used by custom written
coroutines. Likewise, it is intended that custom coroutines use

> the CRTP class template `std::execution::with_awaiable_senders`.

Reviewer 2: ```suggestion
the CRTP class template `std::execution::with_awaitable_senders`.
```

It may be reasonable to adjust the functionality of these components

> instead of definiting the functionality specific to a `lazy<...>`

Reviewer 2: ```suggestion
instead of defining the functionality specific to a `lazy<...>`
```

    work and cleans up upon completion

## `lazy` constructors and assignments

**TODO**

- use guidance from [P1056](https://wg21.link/p1056) and [P2506](https://wg21.link/p2506)

## Result Type For `co_await`

**TODO** turn into text

- status quo based on `execution::as_awaitable(sender)`:
    - Exactly one `set_value_t(T...)` completion
    - `set_value_t()` => `void`
    - `set_value_t(T)` => `decay_t<T>`
    - `set_value_t(T0, T1, ...)` => `tuple<decay_t<T>...>`
- `co_await just_stopped()` could indicate cancellation; supported
   by stdexec.
- `co_await just_error(error)` would be reasonable for completions
    always indicating failure; supported by stdexec.
- the result type is already inconsistent instead of always
    `tuple<decay_t<T>...>` => support multiple `set_value_t`
    completions and produce a `variant<tuple<T...>...>`
- the last option is different from `co_await into_variant<sender>`
    as it treats `set_stopped_t()` and `set_error_t(error)` differently
- it may be reasonable to detect specific shapes of completion
    signatures and turn them into something more coroutine friendly like
    an `optional<T>` or an `expected<T, E>`; alternatively, support

>     corresponding algorithms like `into_optional`, `into_expected`

Reviewer 2: I am default-suspicious of this design direction, but I'll listen to arguments. I don't think the default coroutine type should perform very much magic beyond the surprise-reducing scheduler affinity.

Author: I agree that I'm not fond of necessarily rewriting results. I deliberately list two alternatives which are in the same space to make it obvious that there actually isn't The One Rewrite, either. However, if the rewrite doesn't go into the coroutine itself, I _think_ the proposal should probably contain some algorithms to make the results more coroutine-friend. For example, if `queue.async_pop()` use the completion signatures `set_value_t(T)` and `set_value_t()` to indicate successfully popping an element and that the queue was closed, it may be reasonable to turn that into an `optional<T>` using `into_optional(queue.async_pop())`. Designs of similar interfaces would hopefully follow the direction.

        using sender_concept = ex::sender_t;
        using completion_signatures = ex::completion_signatures<
            ex::set_stopped_t(), Completions...
        >;

        template <ex::receiver R>
        struct state {
            using operation_state_concept = ex::operation_state_t;
            std::decay_t<R> r;
            void start() & noexcept { ex::set_stopped(std::move(r)); }
        };

        template <ex::receiver R>
        state<R> connect(R&& r) { return { std::forward<R>(r) }; }
    };

    ex::lazy<void> fun() {
                                                          // co_await result type
        co_await ex::just();                              // void
        int i = co_await ex::just(0);                     // int
        auto[i, b, c] = co_await ex::just(0, true, 'c');  // tuple<int, bool, char>
        auto v = co_await custom<ex::set_value_t(), ex::set_value_t(int)>;
                                          // variant<tuple<>, tuple<int>>
    }

## Scheduler Affinity

**TODO** turn into text

**TODO** confirm whether it is possible have an inline scheduler which can
    immediately start the operation (or if corresponding detection can go

>     into `continues_on`)

Reviewer 2: `unifex::inline_scheduler`, its Folly equivalent, and the analogous executor-like things in rsys are all bug farms. If the "inline scheduler" you're imagining here is "inline" the way `unifex::inline_scheduler` is "inline" then I'm strongly against having it be a default for the standard coroutine type.

Author: OK. I'll still need to understand (for myself) whether an `inline_scheduler` can actually work (it would just immediately complete in `start()` and I think it works but I haven't tried, yet): it may be a way to explicitly opt out of scheduler affinity.

Also, if the scheduler is mandatory for a receiver connected to a coroutine task, something is needed when there is no scheduler around. For example, I want to `spawn(token, task)` somehow. I think the unifex scope has `scope.spawn_on(task, scheduler)` to make it a bit easier to provide a scheduler. However, I still need some scheduler. I understand that it probably shouldn't be an `inline_scheduler`.

BTW, looking at the ["Spawning work from within a task"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3149r8.html#spawning-work-from-within-a-task) is in that space - it doesn't use any [coroutine] task, though. I think it would be something akin to

    ex::spawn(ex::on(co_await ex::read_env(ex::get_scheduler), std::move(print_sender)), token);

- Resume the coroutine on the same scheduler even if the sender completes
    on a different scheduler: `await_transform(sender)` returns
    `continues_on(as_awaitable(sender), scheduler)`.
- The used `scheduler` is determined using `get_scheduler(env)` where
    `env` is the environment of the task's receiver => the `scheduler`
    needs to be type-erased in some form
- What should happen if there is no `get_scheduler(env)`? Require a
    scheduler (unifex), use an inline scheduler (stdexec), or something

>     else?

Reviewer 2: My ranked preferences are:
 1. require a scheduler, or
 2. something else.

I think "use an inline scheduler" would be a disaster.

Author: I imagine the open questions to become [potential] votes for LEWG to make the wro^Wappropriate decision. There should also be more questions which I'll add: do we want support to report an error without throwing? If so, using any of the proposed approaches? etc.

I will clarify in the actual design discussion that using an inline scheduler is a bad choice.

    a scheduler. In that case an inline scheduler is used. If an inline
    scheduler is used there is the possibility of stack overflow.
- It is possible to `co_await just_error(e)` and `co_await just_stopped()`,
    i.e., the sender isn't required to have a `set_value_t` completion.

The `exec::task<T, C>` also provides a _context_ `C`. An object of
this type becomes the environment for receivers `connect()`ed to
`co_await`ed senders. The default context provides access to the
task's scheduler. In addition an `in_place_stop_token` is provides
which forwards the stop requests from the environment of the receiver
which is connected to the task.

Like the unifex task `exec::task<T, C>` doesn't provide any allocator
support. When creating a task there are two allocations and an
additional allocation for each `co_await`.

# Objectives

Also see [sender/receiver issue 241](https://github.com/cplusplus/sender-receiver/issues/241).

Based on the prior work and discussions around corresponding coroutine
support there a number of required or desired features (listed in
no particular order):

1. A coroutine task needs to be awaiter/awaitable friendly, i.e., it

>     should be possibly to `co_await` awaitables. While that seems

Reviewer: Should we make explicit that this refers to awaitables of all user-supplied types that speak the protocol, and not just `std::lazy` awaitables?

Author: Yes, that clarification is useful. I amended that point.

    obvious, it is possible to create an `await_transform` which
    is deleted for awaiters.
2. When composing sender algorithms without using a coroutine it
    is common to adapt the results using suitable algorithms and
    the completions for sender algorithms are designed accordingly.
    On the other hand, when awaiting senders in a coroutine it may
    be annoying considered having to transform the result into a
    shape which is friendly to a coroutine use. Thus, it may be
    reasonable to support rewriting certain shapes of completion

>     signatures into something different to make the use of senders

Reviewer: Can this paragraph be made more specific with a toy example? Are you just trying to foreshadow "Result Type For `co_await`"? If so, please reference it explicitly.

Author: I want to keep these objectives relatively terse - they are intended primarily as motivating the design discussed later. I have added a forward reference to the relevant section ("result of co_await") and added more discussion and an example to that section. I have also added example code (`demo-into_optional.cpp`).

    set of queries. Normally, queries on an environment can be
    forwarded to the environment of a `connect()`ed receiver.
    Since the coroutine types are determined before the coroutine's
    receiver is known and the queries themselves don't specify a
    result type that isn't possible when a coroutine provides a
    receiver to a sender in a `co_await` expression. It should still
    be possible to provide a user-customisable environment from the
    receiver used by `co_await` expressions. One aspect of this
    environment is to forward stop requests to `co_await`ed child
    operations. Another is possibly changing the scheduler to be
    used when a child operation queries `get_scheduler` from the
    receiver's environment. Also, in non-asynchronous code it is
    quite common to pass some form of context implicitly using
    thread local storage. In an asynchronous world the such contexts
    could be forwarded using the environment.
7. The coroutine should be able to indicate that it was cancelled,
    i.e., to get `set_stopped()` called on the task's receiver.
    `std::execution::with_awaitable_senders` already provided this

>     ability senders being `co_await`ed but that doesn't necessarily

Reviewer: Also, how can a parent coroutine tell if a child was cancelled? Per below, `co_await child()` will short-circuit when the child does `set_stopped()`. However, it can be necessary to handle cancellation explicitly.

If you refer to `co_await std::co_await_result(child())` below, one option would be to have cancellation be represented by a reserved exception type stored in the `exception_ptr` of the result.

If there are protests about "cancellation not being an error", then 
 - `std::result` could potentially have a tri-state API of value / exception / cancellation. I'm not a fan, because that really hurts normal `std::result` usability.
 -  A better direction would be to have `co_await co_nullopt_if_stopped(child())` which returns `std::optional`. Then, in the rare case where you have to handle both cancellation & exception, you write `co_await co_nullopt_if_stopped(co_await_result(child()))` and peel back both wrappers.

Reviewer 2: You can capture the stopped signal with a sender algorithm; two simple choices would be [`let_stopped`](https://eel.is/c++draft/exec.let) for a fully-general transformation or [`upon_stopped`](https://eel.is/c++draft/exec.then) for "transform stopped to a value".

There's also [`stopped_as_optional`](https://eel.is/c++draft/exec.stopped.opt) and [`stopped_as_error`](https://eel.is/c++draft/exec.stopped.err).

Author: It seems I need to clarify that section: what I'm discussing (well, what I'm _trying_ to discuss) here is actually the opposite direction! The comments are about what happens if a child sender completes with `set_stopped()`. I think it is perfectly fine to not resume the coroutine in that situation! If that isn't desirable, rewriting the completion using, e.g., one of the approaches mentioned by Ian is the way to go.

My point was entirely the other way around: It should be possible from within the coroutine to indicate that work was cancelled! The obvious way to do that is, of course, to use

    co_await just_stopped();

    extend to the coroutine implementation.
8. Similar to indicating that a task got cancelled it would be good
    if a task could indicate that an error occurred without throwing
    an exception which escapes from the coroutine.
9. In general a task has to assume that an exception escapes a the
    coroutine implementation. As a result, the task's completion
    signatures need to include `set_error_t(std::exception_ptr)`.
    If it can be indicated to the task that no exception will escape
    the coroutine, this completion signature can be avoided.
10. When many `co_await`ed operations complete synchronously, there
    is a chance for stack overflow. It may be reasonable to have
    the implementation prevent that by using a suitable scheduler
    sometimes.
11. In some situations it can be useful to somehow schedule an
    asychronous clean-up operation which is triggered upon
    coroutine exit.

> 

Reviewer: WDYT about adding point 12 that makes it possible for user-defined types to delegate (by composition) to the `std::lazy` coroutine AND promise type?

The goal would be to cleanly support user-defined coroutine types that forward their awaiter / awaitable interface to `std::lazy`.

https://github.com/facebook/folly/commit/df4cfee9a5fb34609f33e969a72028d3e76c4e75#diff-d37e6fadd609f93f7ab82f70a6e6de773d42496809d6d2b4c2999ca898f2ef3d

Author: If I understand the request correct, I _think_ the current standard text actually already has the relevant facilities and they don't belong into `lazy` at all! In particular, the relevant awaiter/awaitable interface of `lazy` is effectively `std::execution::as_awaitable` which is pretty much what `lazy` uses in its `await_transform`. To get a ready-made coroutine doing all that stuff, a user would derive their promise type from `std::execution::with_awaitable_senders`.

`lazy` uses a custom `await_transform` for two reasons (i.e., it doesn't use `with_awaitable_senders`):

- To get scheduler affinity the sender is wrapped with `continues_on(s, scheduler)`, i.e., the `await_transform()` isn't just calling `as_awaitable(s)` (my current demo implementation also detects an inline scheduler being used and avoids the `continues_on` in that case).
- There are other overloads of `await_transform`, e.g., to facilitate error reporting via `co_await`. I'm not proposing that this direction is pursued but for my demo I wanted to show that it would be possible.

In any case, I don't think supporting user-defined coroutines belongs into `lazy`! I'm supportive of having tools which make the creation of `lazy` effectively a composition of what features are desired, possibly with `lazy` just picking all of them.

Now that I have written a long response saying why I don't want to have that 12th bullet I have come around that I _do_ want a 12th bullet but a different one in the same direction: `lazy` should be composed composed from specified components and a user-defined coroutine could use these components to replicate the respective behaviour to make custom choices for some them.

Reviewer: Thanks, @dietmarkuehl! I like your formulation. My not-too-secret agenda here that it become possible to implement an "immovable `std::lazy`" as a low-effort way of marking which parts of the code are safe from most async lifetime issues.

Of course, as touched on in the Discord, immovability only gives you good lifetime protection if the algorithms taking these tasks are able to take them by-value instead of by-forwarding reference. 

Based on my implementation in `folly::coro`, I'm less sure if the latter point is supportable by a standardized coroutine task, but if there's a cheap fix, I'd say it could be worth it.

Even if we can't improve the standard algorithms to support immovable tasks, having the "roll your own coro" capability is still helpful. E.g. I would be able to port `folly::coro::SafeTask` to this paradigm.

In many cases the receiver can provide access to a scheduler via
the environment query. An example where no scheduler is available
is when starting a task on a [`counting_scope`](https://wg21.link/p3149).
The scope doesn't know about any receivers and, thus, receiver
doesn't support the `get_scheduler` query:

    ex::spawn([]->ex::lazy<void> { co_await ex::just(); }(), token);

Using `spawn()` with work is expected to be quite common, i.e., it
isn't just a theoretical possibility.  The approach used by
[`unifex`](https://github.com/facebookexperimental/libunifex) is
to fail compilation when trying to `connect` a `Task` with a receiver
without a scheduler. The approach taken by
[`stdexec`](https://github.com/NVIDIA/stdexec) is to keep executing
inline in that case. Based on the experience that silently changing
contexts within a coroutine frequently causes bugs it seems failing
to compile is preferrable.

Failing to construct the scheduler used by a coroutine with the
`scheduler` obtained from the receiver is likely an error and should
be addressed by the user appropriately. Failing to compile is seems
to be a reasonable approach in that case, too.

> It should be possible to avoid scheduler affinity explicitly to

Reviewer: `folly::coro`'s design avoids stack overflow and deadlocks related to inline scheduling via the following semantic:
 - A `Task` is lazy, i.e. starts suspended
 - `co_await someCoro()` will execute inline the portion of the the task up to the first suspension (i.e. `co_await` in the coroutine)
 - The continuation after the suspension is scheduled
 - Using an inline scheduler to resume continuation is against contract

Author: Isn't the implication that there is always some rescheduling in that case? I didn't get around experimenting with Folly but I did try a few things out with `unifex` (see `unifex-loop.cpp`): the `unifex` task is always scheduler-aware but it doesn't reschedule work which completes inline. As a result, the [admittedly silly] example actually results in stack overflow. I wonder if Folly avoids that problem.

I still believe that it would be good to somehow allow to avoid scheduler affinity in cases where it is known to be OK and the cost of the rescheduling is a problem. It should be done explicitly and explicitly using an inline scheduler could be such an approach. Is Folly supporting disabling scheduler affinity or a way to not reschedule?

My planned stance on stack overflow is:
 - If you complete synchronously all the time there is probably something wrong to start with - inject something preventing that sometimes!
 - Scheduler affinity normally avoids stack overflow.
 - If necessary (although I don't intend to do so at the moment) stack overflow _could_ be avoided by observing that the completion was synchronous and if there was already a previous one, starting of the awaitable isn't done immediately: instead, a thunk for starting the awaitable is stored, and things return to the previous point where something was started, and the thunk is executed (I did something similar for an implementation of `repeat_until`).

Reviewer: > Isn't the implication that there is always some rescheduling in that case?  

Yes and no. If you take this example here, then it will do everything inline (I checked, no stack overflow in either opt or debug builds), but IIUC stack overflow is prevented by symmetric transfer:

```cpp
#include <iostream>
#include <folly/coro/BlockingWait.h>

int main()
{
    folly::coro::blockingWait([]() -> folly::coro::Task<void> {
        for (int i{}; i < 1000000; ++i) {
            co_await [i]() -> folly::coro::Task<int> {
                co_return i;
            }();
            if (i % 1000 == 0) {
                std::cout << i << "\n" << std::flush;
            }
        }
    }());
    return 0;
}
```

I stepped through the above in debug mode and when the inner task `co_return`s, it resumes into the outer task and destructs the awaiter & inner task, and then continue execution at the `if`. So, IIUC the magical thing here is that [symmetric transfer](https://lewissbaker.github.io/2020/05/11/understanding_symmetric_transfer) makes it so that resuming the outer coro doesn't add a stack frame. I don't believe the above thing goes through the scheduler at all.

Of course, @lewisbaker is the expert on all this, so I could be getting the explanation wrong, but these are facts:
 - When stepping through the above inner loop, no scheduler is involved, just the coros suspending & resuming as you'd expect
 - There is no stack overflow

Reviewer: > somehow allow to avoid scheduler affinity in cases where it is known to be OK and the cost of the rescheduling is a problem

I know of a few things:
 - The "execute inline up to the first suspend" behavior I documented above, which is what makes the above loop demo fast.
 - The ability to `scheduleOn` tasks, which can (but really shouldn't be, for the deadlock reasons discussed above!) used with an inline scheduler.
 - The ability to `startInlineUnsafe`, which you again really really shouldn't do in production code.

Do you have a plausible scenario in mind where the cost of rescheduling is a problem under the "execute inline up to the first suspend" model?

Author: I'm just the person writing the proposal! I'm not the one with experience with that stuff. The fact that there are multiple strategies to avoid rescheduling does imply to me that it is seen as a [potential?] problem. I also suspect that without an opt-out facility people will vote down the proposal on the grounds that it is perceived to introduce unacceptable overhead. I'm sadly not in the position to provide any evidence one way or the other.

I entirely agree that simply running on the completing context is problematic. Based on the discussions I outlined three immediate problems and I hope that is sufficient have scheduler affinity be the default. Explicitly opting out is kind of always possible by providing a form of inline scheduler and the only real question is whether that should provided by the library or not. If the library provides the facility, it can be made such that ability to use scheduler affinity doesn't affect the performance if a user chooses to opt out.

case the error result may be conditionally produced, i.e., it is
possible to complete with an error sometimes and to produce a value
at other times. That could allow a pattern (using `co_yield` for the
potential error return):

    auto value = co_yield when_error(co_await into_expected(sender));

The subexpression `into_expected(sender)` could turn the `set_value_t`
and `set_error_t` into a suitable `std::expected<V, std::variant<E...>>`
always reported using a `set_value_t` completion. The corresponding
`std::expected` becomes the result of the `co_await`. Using `co_yield`
with `when_error(e)` where `e` is an expected can then either produce
`e.value()` as the reuslt of the `co_yield` expression or it can
result in the coroutine completing with the error from `e.error()`.
Using this approach produces a fairly compact approach to propagating
the error retaining the type and without using exceptions.

## Avoiding Stack Overflow

**TODO** turn into text: avoiding stack overflow

- When always using scheduler affinity with a scheduler which doesn't
    immediately return the potential for stack overflows is avoided
- Using an inline scheduler or avoiding the scheduler can expose

>     coroutines to stack overflows

Reviewer: And deadlock, when inline scheduling is mixed with concurrent execution (per the Discord discussion)

Author: This section still needs to be written - these are just notes of what I want to talk about. I added a note on dead lock (and will probably flesh out that section soonish).

- **TODO** add the various other outstanding questions

# Implementation

**TODO** the implementation should be transferred to a Beman project
and equipped with the relevant build, test, examples, and [ideally]
documentation.

An implementation of `lazy` as proposed in this document is available
from [`beman::task26`](https://github.com/maikel/task). This
implementation hasn't received much use, yet, as it is fairly new. It
is setup to be buildable and provides some examples as a starting
point for experimentation.

Coroutine tasks very similar although not identical to the one
proposed are used in multiple projects. In particular, there are
three implementations in wide use:

- [`Folly::Task`](https://github.com/facebook/folly/blob/main/folly/coro/Task.h)
- [`unifex::Task`](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/task.hpp)
- [`stdexec::task`](https://github.com/NVIDIA/stdexec/blob/main/include/exec/task.hpp)

The first one
([`Folly::Task`](https://github.com/facebook/folly/blob/main/folly/coro/Task.h))
isn't based on sender/receiver. Usage experience from all three
have influenced the design of `lazy`.

# Acknowledgements

> We would like to thank Ian Peterson and Alexey [??? TODO] for comments on drafts

Reviewer 2: ```suggestion
We would like to thank Ian Petersen and Alexey Spiridonov for comments on drafts
```

of completions each with a different desirable transformation. If
these are all absorbed into `as_awaitable` it is likely fairly hard
to reason what exact result is returned. Also, there are likely
different options of how a result could be transformed: `into_optional`
is just one example. It could be preferable to turn the two results
into an `std::expected` instead. However, there should probably be
some transformation algorithms like `into_optional`, `into_expected`,
etc.  similar to `into_variant`.

## Scheduler Affinity

Coroutines look very similar to synchronous code with a few
`co`-keywords sprinkled over the code. When reading such code the
expectation is typically that all code executes on the same context
despite some `co_await` expressions using senders which may explicitly
change the scheduler. There are various issues when using `co_await`
naïvely:

* Users may expect that work continues on the same context where it
    was started. If the coroutine simply resumes when the `co_await`ed
    senders calls a completion function code may execute some lengthy
    operation on a context which is expected to keep a UI responsive
    or which is meant to deal with I/O.

> * Conversely, running a loop `co_await`ing some work may be seens as

Reviewer 3: ```suggestion
* Conversely, running a loop `co_await`ing some work may be seen as
```

The allocator used for the coroutine frame should also be used for
any other allocators needed for the coroutine itself, e.g., when
type erasing something needed for its operation (although in most
cases a small object optimization would be preferable and sufficient).
Also, the allocator should be made available to child operations
via the respective receiver's environment using the `get_allocator`
query. The arguments passed to the coroutine are also available to
the constructor of the promise type (if there is a matching on) and
the allocator can be obtained from there:

    struct allocator_aware_context {
        using allocator_type = pmr::polymorphic_allocator<std::byte>;
    };
    fixed_resource<2048> resource;

    ex::sync_wait([](auto&&, auto* resource)
            -> ex::lazy<void, allocator_aware_context> {
        auto alloc = co_await ex::read_env(ex::get_allocator);
        use(alloc);
    }(allocator_arg, &resource));

## Environment Support

When `co_await`ing child operations these may want to access an
environment. Ideally, the coroutine would expose the environment
from the receiver it gets `connect`ed to. Doing so isn't directly

> possible because the coroutine types doesn't know about receiver

Reviewer 3: ```suggestion
possible because the coroutine types doesn't know about the receiver
```

## Avoiding Stack Overflow

It is easy to use a coroutine to accidentally create a stack overflow
because loops don't really execute like loops. For example, a
coroutine like this can easily result in a stack overflow:

    ex::sync_wait(ex::write_env(
        [] -> ex::lazy<void> {
            for (int i{}; i < 1000000; ++i)
                co_await ex::just(i);
        }(),
        ex::make_env(ex::get_scheduler, ex::inline_scheduler{})
    ));
    
The reason this innocent looking code creates a stack overflow is
that the use of `co_await` results in some function calls to suspend
the coroutine and then further function calls to resume the coroutine
(for a proper explanation see, e.g., Lewis Baker's
[Understanding Symmetric Transfer](https://lewissbaker.github.io/2020/05/11/understanding_symmetric_transfer)).
As a result, the stack grows with each iteration of the loop until
it eventually overflows.

With senders it is also not possible to use symmetric transfer to
combat the problem: to achieve the full generality and composing
senders, there are still multiple function calls used, e.g., when
producing the completion signal. Using `get_completion_behavior`

> from the propos [A sender query for completion behaviour](https://wg21.link/P3206)

Reviewer 3: ```suggestion
from the proposal [A sender query for completion behaviour](https://wg21.link/P3206)
```
